rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+        sequence
+        ([`tf.SequenceExample`](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/core/example/example.proto#L298))
+        or structured data (e.g.
+        [Protocol Buffers](https://developers.google.com/protocol-buffers) or
+        [Apache Avro](https://avro.apache.org/)).
+    *   all TFX components can understand and can support their own unique use
+        cases with.
+*   To define an I/O abstraction layer that produces the above in-memory
+    representation from supported physical storage formats, while hiding TFX’s
+    choice of such storage formats from TFX users.
+*   To define a bridge from the above in-memory representation to TF feedables
+    (i.e. Tensors and certain
+    [CompositeTensors](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/python/framework/composite_tensor.py#L1)).
+
+# Motivation
+
+## Fragmented in-memory data representations across TFX
+
+A TFX component may use its input data (data generated by ExampleGen) in two
+ways:
+
+*   it may need to understand the data and conduct analysis. Usually this
+    happens in [Apache Beam](https://beam.apache.org/), and does not involve a
+    TF Model. For example:
+    *   [TFDV](https://github.com/tensorflow/data-validation) and
+        [TFT](https://github.com/tensorflow/transform) compute some statistics
+        over the data.
+    *   [TFMA](https://github.com/tensorflow/model-analysis) may slice the
+        dataset by certain columns in the data.
+*   it may feed the data to TensorFlow. Note that feeding TF alone may not
+    require understanding the data. For example, TFMA may feed a TF model with
+    serialized tf.Example, which may be the raw form of the data.
+
+Currently, each TFX component has its own in-memory data representation to cover
+the two use cases by a different approach:
+
+|    | TFDV | TFT  | TFMA | BulkInference |
+|:---| :--- | :--- | :--- | :------------ |
+| In-memory data representation | Arrow RecordBatches | Dict[str, np.ndarray] | str (raw data records), Dict[str, np.ndarray] | str (raw data records) |
+| Understand the data and conduct analysis | input data is encoded losslessly as RecordBatches. | the in-mem representation may be lossy. | Relies on the model’s input layer, and the format is Dict[str, np.ndarray]. | N/A |
+| Feed TF | N/A | the in-mem representation is TF feedable. | Feed “raw data” to the model. | Feed “raw data” to the model |
+
+This has created many issues:
+
+*   Users of individual components need to adapt their data (if input format not
+    already supported) to each component they want to use.
+*   Individual components rely on unenforceable assumptions on how to interpret
+    the input data consistently.
+*   The complexity of adding new logical data representations (for example,
```

Comment:
It seems "parse the data" here implies a record based I/O abstraction?

Unfortunately in TFX, parsing has to be considered together with I/O because there might not be a record serialization of the data. And the TF Ops that do I/O are tf.Dataset, so unless we bridge the gap between beam's PSource and tf.Dataset, we can't easily use TF OPs to do parsing.

The co-existence of beam's PSource  and tf.Dataset in TFX and their incompatibility is one  (but not the only one) of the factors led to this design -- if I can't use TF for I/O + parsing, why using a TF data structure?

Reply:
> And the TF Ops that do I/O are tf.Dataset, so unless we bridge the gap between beam's PSource and tf.Dataset, we can't easily use TF OPs to do parsing.

How difficult would that be? There are a lot of custom `tf.data.Dataset` adapters in tensorflow/io, so there's some precedent here. Not all of them require a record-based serialization.

I'm not familiar with Beam's extension points, but other similar systems provide a way for users to provide a "mapper" that controls iteration across an individual partition (e.g. Spark's [`RDD.mapPartitions()`](https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.rdd.RDD@mapPartitions[U](f:Iterator[T]=>Iterator[U],preservesPartitioning:Boolean)(implicitevidence$6:scala.reflect.ClassTag[U]):org.apache.spark.rdd.RDD[U])).

For the TFX components that *are* passing their input directly into TF, there's probably some value in optimizing the path to tensors, instead of indirecting via RecordBatch.

/cc @katsiapis @yongtang @jsimsa